### DISCUSSION TITLE
How to model inheritance/polymorphism

### DISCUSSION THREAD

**Discussion by thmxv** [diesel-rs/diesel#3583] [Q&A] [answered]
Hi,

I am migrating a python/sqlalchemy/alembic project to rust. In my existing database (SQLite), most of the data has properties in common and this was modeled using inheritance ("polymorphic_identity" in SQLAlchemy). Do any of you know if this kind of relations is possible with diesel.rs and if I will be able to reuse the same database files.

The current tables looks like this:

Table "object" (the base other tables inherits from):
| id | object_type | date_time_inserted      | date_time_modified    |
|----|----------------|------------------------------|-----------------------------|
| 1  | group         | 20017-11-27 21:10:46 | 20017-11-27 21:10:46 |
| 2  | element      | 20017-11-27 21:10:46 | 20017-11-27 21:10:46 |
| 3  | element      | 20017-11-27 21:10:46 | 20017-11-27 21:10:46 |
| ... | ...                | ...                                 | ...                                 |
| 65 | group         | 20017-11-27 21:10:46 | 20017-11-27 21:10:46 |
| ... | ...                | ...                                 | ...                                 |

Table "group" ( a table that inherits from "object"):
| id  | name    |
|-----|-----------|
| 1   | group 1  |
| 65 | group 2 |
| ... | ...           |

Table "element" (another table that inherits from "object"):
| id | group_id |
|----|------------|
| 2  | 1            |
| 3  | 1            |
| ... | ...           |

I also have a 1 to many relation that use another table "attached_files" and the object id as foreign key.

Thanks in advance for the help. I read the doc and experimented with diesel by following the guides but can't quite figure this out on my own.

**Comment by weiznich** [comment, marked as answer]
This pattern is not supported by diesel out of the box, because it's considered bad database design. That written you can make it work by using some manual trait implementations like the following one:

```rust
use diesel::prelude::*;
use diesel::sql_types::{Integer, Nullable, Text, Timestamp};

table! {
    object(id) {
        id -> Integer,
        object_type -> Text,
        date_time_inserted -> Timestamp,
        date_time_modified -> Timestamp,
    }
}

table! {
    group(id) {
        id -> Integer,
        name -> Text,
    }
}

table! {
    element(id) {
        id -> Integer,
        group_id -> Integer,
    }
}

allow_tables_to_appear_in_same_query!(object, group, element);
joinable!(object -> group(id));
joinable!(object -> element(id));

#[derive(Debug)]
struct Group {
    id: i32,
    name: String,
    date_time_modified: String,
    date_time_inserted: String,
}

#[derive(Debug)]
struct Element {
    id: i32,
    group_id: i32,
    date_time_inserted: String,
    date_time_modified: String,
}

#[derive(Debug)]
enum Object {
    Group(Group),
    Element(Element),
}

impl
    Queryable<
        (
            Integer,
            Text,
            Timestamp,
            Timestamp,
            Nullable<Text>,
            Nullable<Integer>,
        ),
        diesel::sqlite::Sqlite,
    > for Object
{
    type Row = (i32, String, String, String, Option<String>, Option<i32>);

    fn build(row: Self::Row) -> diesel::deserialize::Result<Self> {
        let (id, tpe, inserted, modified, name, group_id) = row;
        match &tpe as &str {
            "group" => Ok(Self::Group(Group {
                id,
                name: name.ok_or_else(|| String::from("name field not set"))?,
                date_time_modified: modified,
                date_time_inserted: inserted,
            })),
            "element" => Ok(Self::Element(Element {
                id,
                group_id: group_id.ok_or_else(|| String::from("group_id field not set"))?,
                date_time_modified: modified,
                date_time_inserted: inserted,
            })),
            _ => Err(String::from("Invalid variant").into()),
        }
    }
}

fn main() {
    let mut conn = SqliteConnection::establish(":memory:").unwrap();

    let r = object::table
        .left_join(group::table)
        .left_join(element::table)
        .select((
            object::id,
            object::object_type,
            object::date_time_inserted,
            object::date_time_modified,
            group::name.nullable(),
            element::group_id.nullable(),
        ))
        .load::<Object>(&mut conn);
}

```

> I also have a 1 to many relation that use another table "attached_files" and the object id as foreign key.

Consider to checkout the ["Relations guide"](http://diesel.rs/guides/relations.html) on our web page for that.

**Comment by thmxv** [comment]
Thanks a lot, that is off tremendous help.

I think I can figure out the "attached_files" relation on my own as it is much more common and covered by the documentation. I made this remark because the foreign key for the attached_files is the object id, so remodeling the DB, to avoid this pattern by adding the columns from the "object" table directly into all the tables that were inheriting from "object" and removing this "object" table, would allow id of "element" and of "group" to be difference auto-increment sequences and this would break the "attached_files" relationship.

I never use the discriminator column "object_type" and always query the child table (joining the object table to access insertion/modification timestamp). What schema would you recommend to use to avoid this design anti-pattern?
- Removing the "object" table and duplicating its columns in the child table (this would require using a single auto-increment index for multiple table primary keys).
- Or removing the discriminator column "object_type" from the object table and adding a "object_id" foreign key to the child tables?

**Reply by weiznich** [reply]
I would probably either just duplicate the columns as the tables are different entities or chose the following structure:

```rust
table! {
    object (id) {
         id -> Integer,
         date_time_inserted -> Timestamp,
         date_time_modified -> Timestamp,
    }   
}

table! {
    group(id) {
        id -> Integer,
        object_id -> Integer,
        name -> Text,
    }
}

table! {
    element(id) {
        id -> Integer,
        object_id -> Integer,
        group_id -> Integer,
    }
}


joinable!(group -> object (object_id));
joinable!(element -> object (object_id));
joinable!(element -> groub (group_id));
```

With this design you have a shared id for entries in the object table, that could be joined from somewhere else + each element/group has a clear reference to object (and not the other way around)
